pkg/virt-handler/notify-server/info.go

```diff
@@ -0,0 +1,53 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2019 Red Hat, Inc.
+ *
+ */
+package eventsserver
+
+import (
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"kubevirt.io/kubevirt/pkg/handler-launcher-com/notify/info"
+	notifyv1 "kubevirt.io/kubevirt/pkg/handler-launcher-com/notify/v1"
+	"kubevirt.io/kubevirt/pkg/virt-launcher/virtwrap/api"
+)
+
+type InfoServer struct{}
+
+func (i InfoServer) Info(context.Context, *info.NotifyInfoRequest) (*info.NotifyInfoResponse, error) {
+
+	// since this is the first versioned version, we only support the current versions
+	// add older versions as soon as they are supported
+	return &info.NotifyInfoResponse{
```

Comment:
all of the work here with setting versions and verifying compatible versions exist is excellent. This allows us to identify when a compatibility issue exists. 

Eventually we are going to get into a scenario where either the client or server has moved forward to a new version of an object, but it needs to interact with its counterpart that is still only aware of the old version of an object. In this scenario, both the old and new versions of the object have to be supported.

do you have any thoughts on how that would be done?

Reply:
Actually we have 2 scenarios, the client moved to a new version, or the server (for `cmd` virt-launcher is the the server, for `notify` it's virt-handler):
- Client is newer (version n): client sends info request, server reports version n-1. Client needs to deal with n-1
- Server is newer (n): client (n-1) sends info request, server needs to support and report n AND n-1, else client will fail

With the current implementation we can detect the need for supporting an old version. But tbh I didn't think about how to actually implement it, as it's not needed for this PR yet imho.

Reply:
> server is newer (n): client (n-1) sends info request, server needs to support and report n AND n-1, else client will fail

Here's an example I'm unclear about.

Lets say the notify-server supports DomainVersion N and N-1. The notify api is just sending json blobs back and forth, so how does the notify-server know whether the client is sending a DomainJSON structured for version N or N-1?

Reply:
Since we know the difference in the JSON between N and N-1, it should be easy to determine the version by just looking into it.
Another approach would be to create a completely new notify / cmd GRPC version for every incompatible change in one the payloads, instead of tracking versions for all kind of payloads. Then the client would just uses the newest version the server offers and it understands itself, and the server knows automatically which version is used. This is the approach the sidecar hooks use. I wanted to avoid that, in order to not having to implement all rpc calls for every version. But the longer I think about it, the more I come to the conclusion that this is a cleaner and easier solution (with more code, but on the other hand the implementing functions for the different versions of the rpc calls can reuse most of the code).
WDYT?

Reply:
> Another approach would be to create a completely new notify / cmd GRPC version for every incompatible change in one the payloads, instead of tracking versions for all kind of payloads. 

Now that we've looked into the details further, I'm inclined to keep it simple using new GRPC version for every incompatible change. 

If/when we have to bump the version someday, maybe it won't be too big of a deal. Depending on the change, I bet we could refactor the client/server code in a way were the bulk of the work is reused.

Reply:
done